### PR TITLE
[FEATURE BRANCH] Stopped rejection of unsecured requests

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Requests received through an unsecured port have no user info.
Previously this filter considered such requests erroneous.  Now it
just handles them immediately; this is just a stop-gap measure, we
should agree upon and then implement better behavior.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
[KEP] https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
/cc @yue9944882 
/cc @aaron-prindle 
/cc @yliaog 
@mars1024 
/sig api-machinery

File: staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go [hunk 1]
```diff
@@ -48,12 +47,14 @@ func WithRequestManagement(
 		ctx := r.Context()
 		requestInfo, ok := apirequest.RequestInfoFrom(ctx)
 		if !ok {
-			handleError(w, r, fmt.Errorf("no RequestInfo found in context"))
+			klog.Errorf("No RequestInfo found in context %#+v of request %#+v", ctx, r)
```
Comment (yue9944882) [inline staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go]: https://github.com/kubernetes/kubernetes/blob/cd5bc3c383b0359029aa82b18828971622cca210/staging/src/k8s.io/apiserver/pkg/server/deprecated_insecure_serving.go#L88-L91

is the change valid? i see other filters does rejections on failing to find request-info/user from ctx..

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go]: (A) If we think the apiserver will not produce this condition, but it does anyway, this is not an error in the request and so the request should not be rejected as an error.

(B) See https://github.com/kubernetes/kubernetes/pull/81224#issuecomment-521806266 for an example showing that there is indeed no userInfo for a request received through an unsecured port (which was the case in that instance).

Comment (yue9944882) [reply staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go]: > (B) See #81224 (comment) for an example showing that there is indeed no userInfo for a request received through an unsecured port (which was the case in that instance).

is the system completely started when you use kubectl? i.e. is `/healthz` returning `ok`? i remember it happens sometimes when we requests an apiserver not completely launched..

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go]: (A) If we think the apiserver will not produce this condition, but it does anyway, this is not an error in the request and so the request should not be rejected as an error.

(B) I will re-investigate for our information, but it does not matter here.  Our filter should not introduce request rejections in situations where requests are not currently rejected.

**Comment by MikeSpreitzer** [comment]
Confirming the problem, I ran `RUNTIME_CONFIG=flowcontrol.apiserver.k8s.io/v1alpha1=true FEATURE_GATES=RequestManagement=true LOG_LEVEL=7 hack/local-up-cluster.sh` in one shell.  I waited until it printed the following things:
```
Local Kubernetes cluster is running. Press Ctrl-C to shut it down.

Logs:
  /tmp/kube-apiserver.log
  /tmp/kube-controller-manager.log
  
  /tmp/kube-proxy.log
  /tmp/kube-scheduler.log
  /tmp/kubelet.log

To start using your cluster, you can open up another terminal/tab and run:

  export KUBECONFIG=/var/run/kubernetes/admin.kubeconfig
  cluster/kubectl.sh

Alternatively, you can write to the default kubeconfig:

  export KUBERNETES_PROVIDER=local

  cluster/kubectl.sh config set-cluster local --server=https://localhost:6443 --certificate-authority=/var/run/kubernetes/server-ca.crt
  cluster/kubectl.sh config set-credentials myself --client-key=/var/run/kubernetes/client-admin.key --client-certificate=/var/run/kubernetes/client-admin.crt
  cluster/kubectl.sh config set-context local --cluster=local --user=myself
  cluster/kubectl.sh config use-context local
  cluster/kubectl.sh
```

In another shell I did the following.

```
mspreitz@mjs-dev-1:~/go/src/k8s.io/kubernetes$ git rev-parse HEAD
4229aca175cbcf11934bfc832d82380233de22ec

mspreitz@mjs-dev-1:~/go/src/k8s.io/kubernetes$ curl http://localhost:8080/healthz
Internal Server Error: "/healthz"

mspreitz@mjs-dev-1:~/go/src/k8s.io/kubernetes$ curl http://localhost:8080/
Internal Server Error: "/"

mspreitz@mjs-dev-1:~/go/src/k8s.io/kubernetes$ KUBECONFIG=/var/run/kubernetes/admin.kubeconfig kubectl get --raw /healthz
ok
mspreitz@mjs-dev-1:~/go/src/k8s.io/kubernetes$ kubectl get ns fooobar
Error from server (InternalError): an error on the server ("Internal Server Error: \"/api?timeout=32s\"") has prevented the request from succeeding
```

As you can see, even the unsecured requests for `/healthz` and `/` were rejected.  The attempt to use `kubectl` to look at a namespace failed at the API introspection step; here is what was logged by the apiserver:
```
E0820 15:32:22.416167    7851 maxinflight.go:50] no User found in context
I0820 15:32:22.416208    7851 httplog.go:90] GET /api?timeout=32s: (169.416µs) 500
goroutine 14313 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).recordStatus(0xc003046b60, 0x1f4)
        /home/mspreitz/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:217 +0xc8
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader(0xc003046b60, 0x1f4)
        /home/mspreitz/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:196 +0x35
net/http.Error(0x89109a0, 0xc003046b60, 0xc006b70c90, 0x29, 0x1f4)
        /usr/local/go/src/net/http/server.go:2007 +0x213
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.handleError(0x89109a0, 0xc003046b60, 0xc006820c00, 0x888cbc0, 0xc00449c570)
        /home/mspreitz/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/maxinflight.go:49 +0xca
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithRequestManagement.func1(0x89109a0, 0xc003046b60, 0xc006820c00)
        /home/mspreitz/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:56 +0x7a3
net/http.HandlerFunc.ServeHTTP(0xc003e02ba0, 0x89109a0, 0xc003046b60, 0xc006820c00)
        /usr/local/go/src/net/http/server.go:1995 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithWaitGroup.func1(0x89109a0, 0xc003046b60, 0xc006820c00)
        /home/mspreitz/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/waitgroup.go:47 +0xf3
net/http.HandlerFunc.ServeHTTP(0xc003e02bd0, 0x89109a0, 0xc003046b60, 0xc006820c00)
        /usr/local/go/src/net/http/server.go:1995 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithRequestInfo.func1(0x89109a0, 0xc003046b60, 0xc006820b00)
        /home/mspreitz/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:39 +0x2b8
net/http.HandlerFunc.ServeHTTP(0xc003e02c60, 0x89109a0, 0xc003046b60, 0xc006820b00)
        /usr/local/go/src/net/http/server.go:1995 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.WithLogging.func1(0x8914460, 0xc0073d7ea0, 0xc006820a00)
        /home/mspreitz/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:89 +0x29c
net/http.HandlerFunc.ServeHTTP(0xc003dcf800, 0x8914460, 0xc0073d7ea0, 0xc006820a00)
        /usr/local/go/src/net/http/server.go:1995 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.withPanicRecovery.func1(0x8914460, 0xc0073d7ea0, 0xc006820a00)
        /home/mspreitz/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/wrap.go:44 +0x105
net/http.HandlerFunc.ServeHTTP(0xc003dcf820, 0x8914460, 0xc0073d7ea0, 0xc006820a00)
        /usr/local/go/src/net/http/server.go:1995 +0x44
net/http.serverHandler.ServeHTTP(0xc0023b9ba0, 0x8914460, 0xc0073d7ea0, 0xc006820a00)
        /usr/local/go/src/net/http/server.go:2774 +0xa8
net/http.(*conn).serve(0xc001db1a40, 0x8925120, 0xc007130b00)
        /usr/local/go/src/net/http/server.go:1878 +0x851
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:2884 +0x2f4

logging error output: "Internal Server Error: \"/api?timeout=32s\"\n"
 [kubectl/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:57506]
```

**Comment by MikeSpreitzer** [comment]
Please note the difference between request info and user info.  What I demonstrated is a problem with lack of user info.  The code you cited is looking for request info.  Even requests received through an unsecured port have request info.

**Comment by MikeSpreitzer** [comment]
I would be willing to change the reaction to missing request info back to rejection, on the basis that this is what the max-in-flight filter does.  But I think we definitely need the change here for the case of missing user info.

File: staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go (hunk 1 above)
Comment (yliaog) [reply staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go]: I agree we should not introduce new request rejections cases, but should we still do rate limiting when request info or user info is missing? i think we should. WDYT?

Comment (yue9944882) [reply staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go]: i think the error is suggesting sth is wrong w/ the insecure handler chaining, not the req-mgmt handler to blame.. i am slightly leaning on a fix from the master

**Pull request closed without merging**